### PR TITLE
Device de/rehydration flows

### PR DESCRIPTION
**Pull request by stefanceriu**
Fixes vector-im/element-ios/issues/4255 - Device de/rehydration flows.

### CODE REVIEW

File: MatrixKit/Models/Account/MXKAccount.m [hunk 1]
```diff
@@ -862,6 +862,10 @@ -(void)openSessionWithStore:(id<MXStore>)store
         // Launch server sync
         [self launchInitialServerSync];
         
+        if ([self.mxSession.crypto.crossSigning canCrossSign]) {
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccount.m:865]:
```suggestion
        if ([self.mxSession.crypto.crossSigning canCrossSign])
        {
```

File: MatrixKit/Models/Account/MXKAccount.m [hunk 2]
```diff
@@ -1673,6 +1677,39 @@ - (void)launchInitialServerSync
...
+    
+    MXKeyData * keyData = [[MXKeyProvider sharedInstance] requestKeyForDataOfType:MXDehydrationServiceKeyDataType isMandatory:NO expectedKeyType:kRawData];
+    if (keyData == nil)
+    {
+        MXLogWarning(@"[MXKAccount] no key provided for device dehydration");
+    }
+    
+    NSData *key = ((MXRawDataKey*) keyData).key;
+    
+    MXLogDebug(@"[MXKAccount] starting device dehydration");
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccount.m:1699]: Method name

File: MatrixKit/Models/Account/MXKAccount.m [hunk 3]
```diff
@@ -1673,6 +1677,39 @@ - (void)launchInitialServerSync
...
+    [[MXKAccountManager sharedManager].dehydrationService dehydrateDeviceWithMatrixRestClient:mxRestClient crypto:mxSession.crypto dehydrationKey:key success:^(NSString *deviceId) {
+        [MXKAppSettings standardAppSettings].enableDehydration = NO;
+        MXLogDebug(@"[MXKAccount] device successfully dehydrated");
+    } failure:^(NSError *error) {
+        if (retry) {
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccount.m:1704]: Code format
We miss a code format linter in the CI :/

**Review by manuroe — COMMENTED** [review]
LGTM
Please make sure that all logs contain the method name.

File: MatrixKit/Models/Account/MXKAccount.m (hunk 3 above)
Comment (stefanceriu) [reply MatrixKit/Models/Account/MXKAccount.m:1704]: Yup, locally as well but I think that ship has sailed for ObjC

File: MatrixKit/Models/Account/MXKAccount.m [hunk 4]
```diff
@@ -1684,28 +1685,31 @@ - (void)attemptDeviceDehydration
...
     } failure:^(NSError *error) {
-        if (retry) {
+        if (retry)
+        {
             [self attemptDeviceDehydrationWithRetry:NO];
-            MXLogError(@"[MXKAccount] device dehydration failed due to error: %@. Retrying.", error);
-        } else {
-            MXLogError(@"[MXKAccount] device dehydration failed due to error: %@", error);
+            MXLogError(@"[MXKAccount] attemptDeviceDehydrationWithRetry: device dehydration failed due to error: %@. Retrying.", error);
+        } else
```
Comment (manuroe) [inline MatrixKit/Models/Account/MXKAccount.m:1710]:
```suggestion
        } 
        else
```

almost :)

**Review by manuroe — COMMENTED** [review]
One final bit

File: MatrixKit/Models/Account/MXKAccount.m (hunk 4 above)
Comment (stefanceriu) [reply MatrixKit/Models/Account/MXKAccount.m:1710]: Damn it! 😅

File: MatrixKit/Controllers/MXKAuthenticationViewController.h [hunk 5]
```diff
@@ -286,12 +286,26 @@
...
 
 /**
  Display the fallback URL within a webview.
  */
 - (void)showAuthenticationFallBackView;
 
+#pragma mark - Device rehydration
+
+/**
+ Call this method at an appropriate time to attempt rehydrating from an existing backup device
```
Comment (manuroe) [inline MatrixKit/Controllers/MXKAuthenticationViewController.h:303]:
```suggestion
 Call this method at an appropriate time to attempt rehydrating from an existing dehydrated device
```

to keep matrix.org spec wording

**Review by manuroe — COMMENTED** [review]
A single suggestion in comments

**Pull request merged**